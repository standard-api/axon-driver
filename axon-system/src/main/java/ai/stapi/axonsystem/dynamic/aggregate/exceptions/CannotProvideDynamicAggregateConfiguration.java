package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

public class CannotProvideDynamicAggregateConfiguration extends Exception {

  public CannotProvideDynamicAggregateConfiguration(String aggregateType) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.",
            aggregateType
        )
    );
  }

  public CannotProvideDynamicAggregateConfiguration(String aggregateType, Throwable cause) {
    super(
        String.format(
            "Cannot provide Dynamic Aggregate Configuration for '%s', because it does not exists.\n Cause: %s",
            aggregateType,
            cause.getMessage()
        ),
        cause
    );
  }
}
