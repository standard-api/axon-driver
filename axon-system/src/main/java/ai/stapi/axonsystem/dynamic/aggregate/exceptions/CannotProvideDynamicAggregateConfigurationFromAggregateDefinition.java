package ai.stapi.axonsystem.dynamic.aggregate.exceptions;

import ai.stapi.graphsystem.aggregatedefinition.model.CommandHandlerDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;

public class CannotProvideDynamicAggregateConfigurationFromAggregateDefinition
    extends RuntimeException {

  private CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(String becauseMessage) {
    super("Cannot provide DynamicAggregateConfiguration from OperationDefinition, because "
        + becauseMessage);
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseOperationHasNoneOrMoreResources(
      OperationDefinitionDTO operationDefinitionDTO
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "operation definition has none or more resources." +
            "\nOperation definition: " + operationDefinitionDTO
    );
  }

  public static CannotProvideDynamicAggregateConfigurationFromAggregateDefinition becauseCommandHandlerDefinitionHasUnknownCreationalPolicy(
      CommandHandlerDefinitionDTO commandDefinition
  ) {
    return new CannotProvideDynamicAggregateConfigurationFromAggregateDefinition(
        "command handler definition has unknown creational policy." +
            "\nCommand handler definition: " + commandDefinition
    );
  }
}
