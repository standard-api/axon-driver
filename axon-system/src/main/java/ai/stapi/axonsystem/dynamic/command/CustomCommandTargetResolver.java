package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
