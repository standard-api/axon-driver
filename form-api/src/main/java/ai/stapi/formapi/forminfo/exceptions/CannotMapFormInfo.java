package ai.stapi.formapi.forminfo.exceptions;

public class CannotMapFormInfo extends RuntimeException {

    private static final String MSG = "Cannot map form info, because ";

    private CannotMapFormInfo(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotMapFormInfo(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotMapFormInfo becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotMapFormInfo(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }
}
