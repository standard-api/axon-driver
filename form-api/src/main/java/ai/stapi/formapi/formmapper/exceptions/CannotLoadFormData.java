package ai.stapi.formapi.formmapper.exceptions;

public class CannotLoadFormData extends RuntimeException {

    private static final String MSG = "Cannot load data for form, because ";

    private CannotLoadFormData(String becauseMessage) {
        super(MSG + becauseMessage);
    }

    private CannotLoadFormData(String becauseMessage, Throwable cause) {
        super(MSG + becauseMessage, cause);
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            )
        );
    }

    public static CannotLoadFormData becauseThereWasNoCommandHandlerDefinitionForProvidedOperation(
        String operationId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "there was not command handler definition with provided operation.%nOperation name: '%s'",
                operationId
            ),
            cause
        );
    }

    public static CannotLoadFormData becauseAggregateByProvidedResourceIdWasNotFound(
        String resourceId,
        Throwable cause
    ) {
        return new CannotLoadFormData(
            String.format(
                "no aggregate with provided resource id was found.%nResource id: '%s'",
                resourceId
            ),
            cause
        );
    }
}
