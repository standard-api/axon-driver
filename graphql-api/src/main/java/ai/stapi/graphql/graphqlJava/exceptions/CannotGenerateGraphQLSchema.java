package ai.stapi.graphql.graphqlJava.exceptions;

public class CannotGenerateGraphQLSchema extends RuntimeException {

  private CannotGenerateGraphQLSchema(String becauseMessage) {
    super("Cannot generate graphQl schema, because " + becauseMessage);
  }

  public static CannotGenerateGraphQLSchema becauseStructureSchemaFieldHadNoType(
      String typeName,
      String fieldName
  ) {
    return new CannotGenerateGraphQLSchema(
        "structure schema has some field which has no type." +
            "\nType name: " + typeName +
            "\nField name: " + fieldName
    );
  }

  public static CannotGenerateGraphQLSchema becauseOperationReferencedTypeWhichDoesNotExistInStructureSchema(
      String operationName,
      String parameterTypeName
  ) {
    return new CannotGenerateGraphQLSchema(
        "operation has parameter of type which does not exist in structure schema." +
            "\nOperation name: " + operationName +
            "\nParameter Type name: " + parameterTypeName
    );
  }
}
