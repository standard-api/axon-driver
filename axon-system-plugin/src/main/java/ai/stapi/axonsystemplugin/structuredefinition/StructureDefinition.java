package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.axonsystem.graphaggregate.AggregateWithDynamicGraph;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.FixedImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.schema.structuredefinition.StructureDefinitionNormalizer;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.CreationPolicy;
import org.axonframework.spring.stereotype.Aggregate;
import org.springframework.beans.factory.annotation.Autowired;

@Aggregate
public class StructureDefinition extends AggregateWithDynamicGraph<StructureDefinitionId> {

  @CommandHandler
  @CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
  public void handle(
      ImportStructureDefinition command,
      @Autowired DynamicCommandProcessor commandProcessor
  ) {
    this.commandProcessor = commandProcessor;
    this.processCommandDynamically(
        new FixedImportStructureDefinition(
            command.getTargetIdentifier(),
            StructureDefinitionNormalizer.normalize(command.getStructureDefinitionSource())
        ),
        MissingFieldResolvingStrategy.LENIENT
    );
  }

  @EventSourcingHandler
  public void on(StructureDefinitionImported event) {
    this.onAggregateCreated(event);
  }
}