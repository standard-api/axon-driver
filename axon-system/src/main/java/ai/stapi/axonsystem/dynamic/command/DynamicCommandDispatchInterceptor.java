package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
