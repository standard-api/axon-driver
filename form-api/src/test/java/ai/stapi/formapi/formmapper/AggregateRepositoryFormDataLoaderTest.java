package ai.stapi.formapi.formmapper;

import ai.stapi.formapi.formmapper.exceptions.CannotLoadFormData;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionDTO;
import ai.stapi.graphsystem.operationdefinition.model.OperationDefinitionProvider;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionData;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionElementDefinition;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.test.domain.DomainTestCase;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

class AggregateRepositoryFormDataLoaderTest extends DomainTestCase {
    
    @Autowired
    private AggregateRepositoryFormDataLoader aggregateRepositoryFormDataLoader;
    
    @Autowired
    private OperationDefinitionProvider operationDefinitionProvider;
    
    @Test
    void itThrowsExceptionWhenCommandHandlerDefinitionForProvidedOperationDoesntExist() {
        Executable throwable = () -> this.aggregateRepositoryFormDataLoader.load(
            new OperationDefinitionDTO(
                "NotExisting",
                "NotExisting",
                "Irrelevant",
                "Irrelevant",
                "Irrelevant",
                "Irrelevant",
                List.of(),
                false,
                false,
                false,
                List.of()
            ),
            "Irrelevant",
            Map.of()
        );
        this.thenExceptionMessageApproved(CannotLoadFormData.class, throwable);
    }

    @Test
    void itThrowsExceptionWhenAggregateWithProvidedResourceIdDoesntExist() {
        var operationId = "CreateParameters";
        var resourceId = "NonExistingId";
        var operation = this.operationDefinitionProvider.provide(operationId);
        Executable throwable = () -> this.aggregateRepositoryFormDataLoader.load(
            operation,
            resourceId,
            Map.of()
        );
        this.thenExceptionMessageApproved(CannotLoadFormData.class, throwable);
    }

    @Test
    void itCanLoadFormDataForSimpleOperation() {
        var operationId = "CreateParameters";
        var resourceId = "SomeId";
        this.givenCommandIsDispatched(
            new DynamicCommand(
                new UniqueIdentifier(resourceId),
                operationId,
                Map.of(
                    "language", "cz",
                    "implicitRules", "http://some.uri.com"
                )
            )
        );
        var operation = this.operationDefinitionProvider.provide(operationId);
        var actual = this.aggregateRepositoryFormDataLoader.load(
            operation,
            resourceId,
            Map.of()
        );
        this.thenObjectApproved(actual);
    }

    @Test
    void itCanLoadFormDataForComplexCommand() {
        var operationId = "CreateStructureDefinition";
        var resourceId = "SomeId";
        var data = new ObjectMapper().convertValue(
            new RawStructureDefinitionData(
                resourceId,
                "http://some.url/SomeId",
                "draft",
                "Example Description",
                "complex-type",
                false,
                resourceId,
                null,
                new RawStructureDefinitionData.Differential(
                    new ArrayList<>(List.of(
                        new RawStructureDefinitionElementDefinition(
                            "ExampleComplex.exampleField",
                            1,
                            "1",
                            "Example field description",
                            "Example field description",
                            "Example field description",
                            new ArrayList<>(List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            ))
                        )
                    ))
                )
            ),
            new TypeReference<Map<String, Object>>() {
            }
        );
        this.givenCommandIsDispatched(
            new DynamicCommand(
                new UniqueIdentifier(resourceId),
                operationId,
                data
            )
        );
        var operation = this.operationDefinitionProvider.provide(operationId);
        var actual = this.aggregateRepositoryFormDataLoader.load(
            operation,
            resourceId,
            Map.of()
        );
        this.thenObjectApproved(actual);
    }
}