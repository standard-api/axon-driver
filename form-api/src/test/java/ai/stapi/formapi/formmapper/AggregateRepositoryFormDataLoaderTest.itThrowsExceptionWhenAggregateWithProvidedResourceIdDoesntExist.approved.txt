Cannot load data for form, because no aggregate with provided resource id was found.
Resource id: 'NonExistingId'